python_modules/dagster/dagster/_core/execution/plan/execute_step.py

```diff
@@ -703,12 +714,21 @@ def _store_output(
     manager_materializations = []
     manager_metadata: Dict[str, MetadataValue] = {}
 
+    used_io_manager = True
+
     # don't store asset check outputs or asset observation outputs
     step_output = step_context.step.step_output_named(step_output_handle.output_name)
     asset_key = step_output.properties.asset_key
-    if step_output.properties.asset_check_key or (
-        step_context.output_observes_source_asset(step_output_handle.output_name)
+    if (
+        step_output.properties.asset_check_key
+        or (step_context.output_observes_source_asset(step_output_handle.output_name))
+        or (
+            output_context.has_asset_key
+            and output_context.dagster_type.is_nothing
+            and output.value is None
+        )
     ):
+        used_io_manager = False
```

Comment:
theres gotta be a better way to factor out the output handling so this control flow stuff is more clear